client/lib/service-worker/service-worker.js

```diff
@@ -100,3 +116,156 @@ self.addEventListener( 'message', function( event ) {
 			break;
 	}
 } );
+
+self.addEventListener( 'fetch', function( event ) {
+	const request = event.request;
+
+	if ( request.method !== 'GET' ) {
+		return;
+	}
+
+	// HTML Pages, fetch from the server and fallback to the Offline page
+	if ( request.mode === 'navigate' ) {
+		const previousHash = currentAssetsHash;
+		if ( navigator.onLine ) {
+			event.respondWith(
+				fetchNetworkFirst( request, OFFLINE_CALYPSO_PAGE ).then( function( response ) {
+					// Detect if the loaded page is calypso or not
+					// TODO: update nginx to reverse proxy this header
+					if ( response.headers.get( 'x-powered-by' ) === 'Express' ) {
+						// Let's check that assets have not changed before serving the page
+						// Fetching assets.json on each page load should be relatively lightweight as
```

Comment:
it may be light in terms of bytes, but it could still be very slow in terms of latency, especially on flaky mobile networks. If we're only pushing 5kb, but it takes 10s to get the first byte, does that change how this should work?